klippy/configfile.py

```diff
@@ -331,17 +332,29 @@ def get_status(self, eventtime):
         return {'config': self.status_raw_config,
                 'settings': self.status_settings,
                 'warnings': self.status_warnings,
-                'save_config_pending': self.save_config_pending}
+                'save_config_pending': self.save_config_pending,
+                'save_config_pending_items': self.status_save_pending}
     # Autosave functions
     def set(self, section, option, value):
         if not self.autosave.fileconfig.has_section(section):
             self.autosave.fileconfig.add_section(section)
         svalue = str(value)
         self.autosave.fileconfig.set(section, option, svalue)
+        pending = dict(self.status_save_pending)
+        if not section in pending:
+            pending[section] = {}
+        else:
+            pending[section] = dict(pending[section])
+        pending[section][option] = svalue
+        self.status_save_pending = pending
         self.save_config_pending = True
         logging.info("save_config: set [%s] %s = %s", section, option, svalue)
     def remove_section(self, section):
         self.autosave.fileconfig.remove_section(section)
+        if section in self.status_save_pending:
```

Comment:
This is correct in that it would remove items in the "status_save_pending" dict, however a removed item not added in this session won't be indicated.  For example, if a user saved a mesh profile in this session and then subsequently removed it then it is correctly handled...however if the profile was saved in a previous section and it was removed there is no indication that the removal is pending.

Reply:
I think I understand what you mean, but let me phrase it my way to see whether we're on the same line: the status_save_pending is not really reflecting when an item is removed as the removal of an item is not flagged explicitly in the dict as a removal, just the absence of the value.

An alternative could be to add or keep the entry in the dict, but give it a None value.  Would that be better?

Reply:
You could assign the section a `None` value if its removed, however it might require some additional tracking to make sure it was there when the session started.  Another option would be to report the autosave data as a dict.  Then users would have a view of exactly what the "saved" portion of the configuration will be when they issue `SAVE_CONFIG`.  I'd be curious as to what Kevin and Meteyou would think is the best approach.

Reply:
I guess I'm still somehow missing your exact point.  We seem to be on the same line on what the goal should be: having a clear view on the current settings AND how those differ from the already saved settings (which is also the settings at start of the session as a SAVE_CONFIG triggers a restart).

My main goal would be that a GUI (Ex. Mainsail) or a set of macros should be able to easily figure out which config values have not yet been saved to reflect that to the user or - which is what I use it for - store those in some other way that does not require a restart and then restore them on the next startup to avoid they get lost.  Why would they get lost?  Because after ex. z-adjust at the beginning of a print, you cannot do a SAVE_CONFIG without killing the print and I have totally separate system that monitors the printer and shuts down power when it is cold for 15 minutes.  So, if power is taken off 15 minutes after the print ended, my z-offset is gone - were it not for my macros saving stuff in saved variables and then at startup restoring all that.  Question then is: why do I need to make such tricky macros for this?

In general I guess the problem is running deeper inside Klipper.  There are settings which for very practical reasons have to be applied at the start of a session (pin-configurations and other hardware-related stuff).  However, there are numerous settings which can be dynamically modified at run-time ((additional) z-offset, bed profiles, max.speeds/accel, ...) yet for some reason also these values can only be saved in combination with a restart.  Ultimately I think it would make more sense to have 2 separate types of config, one for stuff that cannot change dynamically (and might even be read-only, like the pin-config) and one for stuff that is dynamic and could hence be saved at any point.  Even then an "original values" vs "current values" view via 2 dicts might be useful, but the need for it would already be much lower.

Reply:
Having to restart to save various settings is indeed annoying.
Even a "save without kill" would be a definite improvement

Reply:
@Arksine , I'm having problems figuring out what was part of the original config or not.  Something like:
```
    def remove_section(self, section):
        if self.autosave.fileconfig.has_section(section):
            self.autosave.fileconfig.remove_section(section)
            pending = dict(self.status_save_pending)
            pending[section] = None
            self.status_save_pending = pending
            self.save_config_pending = True
        elif section in self.status_save_pending and self.status_save_pending[section] is not None:
            pending = dict(self.status_save_pending)
            del pending[section]
            self.status_save_pending = pending
            self.save_config_pending = True
```
does not seem to work because autosave.fileconfig is reflecting current status, not the original unchanged config.  Am I overlooking something or do I really have to make a copy of the initial config (RawConfigParser).  The latter seems pretty heavy and I wonder whether knowing the distinction between those two is worth it?

Reply:
Having given it some thought I'm not sure that the distinction is necessary.  If `self.status_save_pending` has a section set to `None` then a front end or macro will conclude that the section will not be saved when `SAVE_CONFIG` is run.   I think the result it the same regardless of whether or not the section was initially loaded.

Reply:
Ok.  I do still want to check the effect of these None values further.  If that behaves nicely, I'll update the pull-request with this change.

Reply:
It required some extra fixes to get through the tests - one real issue due to the "None" value now possible for a section which I had apparently not covered in my usage the past days, but luckily was exposed by the tests.